File: kb-rag-web/src/api/webSource.ts
// Author: owlzhangfq@gmail.com
import { apiDelete, apiGet, apiPost, apiPut } from './request';
import type { PageResult, RegisterWebSourceRequest, WebSourceEntry } from './types';

/**
 * POST /api/v1/kb/{kbId}/web-sources (M12-CONTRACTS.md section 3.4): registers a page URL and
 * runs its first fetch inline, so the returned entry already carries the outcome -- FAILED here
 * means the registration stuck but the page could not be fetched (the row keeps the error).
 */
export function registerWebSource(kbId: string, payload: RegisterWebSourceRequest): Promise<WebSourceEntry> {
  return apiPost<WebSourceEntry>(`/kb/${kbId}/web-sources`, payload);
}

/** GET /api/v1/kb/{kbId}/web-sources (M12-CONTRACTS.md section 3.4), most recently registered first. */
export function listWebSources(kbId: string, page = 1, size = 20): Promise<PageResult<WebSourceEntry>> {
  return apiGet<PageResult<WebSourceEntry>>(`/kb/${kbId}/web-sources`, { page, size });
}

/**
 * POST /api/v1/web-sources/{sourceId}/sync (M12-CONTRACTS.md section 3.4): one on-demand sync.
 * Never rejects for a fetch problem -- the outcome (SUCCESS/UNCHANGED/SKIPPED/FAILED) rides on
 * the returned entry's last_fetch_status, matching what the scheduled pass would have recorded.
 */
export function syncWebSource(sourceId: string): Promise<WebSourceEntry> {
  return apiPost<WebSourceEntry>(`/web-sources/${sourceId}/sync`);
}

/** PUT /api/v1/web-sources/{sourceId} (M12-CONTRACTS.md section 3.4): flips the scheduled-sync switch. */
export function updateWebSource(sourceId: string, syncEnabled: boolean): Promise<WebSourceEntry> {
  return apiPut<WebSourceEntry>(`/web-sources/${sourceId}`, { sync_enabled: syncEnabled });
}

/**
 * DELETE /api/v1/web-sources/{sourceId} (M12-CONTRACTS.md section 3.4): removes the registration
 * only -- the document it fed stays in the knowledge base (weak binding).
 */
export function removeWebSource(sourceId: string): Promise<void> {
  return apiDelete<void>(`/web-sources/${sourceId}`);
}

File: kb-rag-server/kb-domain/src/main/java/io/kbrag/domain/entity/WebSource.java
package io.kbrag.domain.entity;

import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableName;
import io.kbrag.domain.enums.WebSourceFetchStatus;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

import java.time.LocalDateTime;

/**
 * A registered web page source, the M12 contract section 1: the binding between a URL and the
 * document its fetches produce, plus the sync switch and the outcome of the last fetch.
 *
 * <p>The binding is deliberately weak. Removing the registration leaves the document alone, and
 * trashing the document leaves the registration alone - the two lifecycles meet only inside a sync
 * pass, which reads both sides and decides what one fetch may do.
 *
 * @author owlzhangfq@gmail.com
 */
@Getter
@Setter
@ToString(callSuper = true)
@TableName("t_kb_web_source")
public class WebSource extends BaseEntity {

    private static final long serialVersionUID = 1L;

    /** Business identifier exposed through the API. */
    @TableField("source_id")
    private String sourceId;

    /** Knowledge base the fetched pages land in. */
    @TableField("kb_id")
    private String kbId;

    /** Registered page address, http or https. */
    @TableField("url")
    private String url;

    /** SHA-256 of the URL; the equality key a VARCHAR(2048) column cannot be. */
    @TableField("url_hash")
    private String urlHash;

    /** Document the fetches feed, {@code null} until the first successful fetch. */
    @TableField("doc_id")
    private String docId;

    /** Derived stable file name the upload chain sees, {@code null} until the first fetch. */
    @TableField("file_name")
    private String fileName;

    /** {@code 1} includes the source in the scheduled sync pass. */
    @TableField("sync_enabled")
    private Integer syncEnabled;

    /** {@code 1} fetches this source through the headless browser and stores the rendered DOM, the M17 contract section 1. */
    @TableField("render_js")
    private Integer renderJs;

    /** SHA-256 of the last fetched body, the unchanged check of an incremental sync. */
    @TableField("last_content_hash")
    private String lastContentHash;

    /** When the last sync attempt ran, success or not. */
    @TableField("last_fetch_at")
    private LocalDateTime lastFetchAt;

    /** Outcome of the last sync attempt. */
    @TableField("last_fetch_status")
    private WebSourceFetchStatus lastFetchStatus;

    /** Why the last sync failed or was skipped, {@code null} on success. */
    @TableField("last_error")
    private String lastError;
}
